src/s3path.jl

```diff
@@ -107,25 +111,35 @@ function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Noth
     path = ()
     isdirectory = true
 
-    tokenized = split(str, "/")
-    bucket = strip(tokenized[3], '/')
-    drive = "s3://$bucket"
+    uri = URI(str)
+    params = queryparams(uri)
+    version = get(params, "versionId", nothing)
+    bucket = uri.host
 
-    if length(tokenized) > 3
+    drive = "s3://$bucket"
+    tokenized = isempty(uri.path) ? nothing : split(lstrip(uri.path, '/'), '/')
+    if tokenized !== nothing
         root = "/"
         # If the last tokenized element is an empty string then we've parsed a directory
         isdirectory = isempty(last(tokenized))
-        path = Tuple(filter!(!isempty, tokenized[4:end]))
+        path = Tuple(filter!(!isempty, tokenized))
     end
 
-    return S3Path(path, root, drive, isdirectory, config, nothing)
+    return S3Path(path, root, drive, isdirectory, config, version)
 end
 
+
+
 function normalize_bucket_name(bucket)
     return strip(startswith(bucket, "s3://") ? bucket : "s3://$bucket", '/')
 end
 
-Base.print(io::IO, fp::S3Path) = print(io, fp.anchor * fp.key)
+function Base.print(io::IO, fp::S3Path)
+    if fp.version === nothing || isempty(fp.version)
+        return print(io, fp.anchor * fp.key)
+    end
+    return print(io, fp.anchor * fp.key * "?versionId=" * fp.version)
```

Comment:
Still making intermediate strings:
```suggestion
    if fp.version === nothing || isempty(fp.version)
        return print(io, fp.anchor, fp.key)
    end
    return print(io, fp.anchor, fp.key, "?versionId=", fp.version)
```

Reply:
Alternatively, why duplicate the logic for printing anchor and key:
```suggestion
    print(io, fp.anchor, fp.key)
    if fp.version === nothing || isempty(fp.version)
        print(io, "?versionId=", fp.version) 
    end
    return nothing
```
This is what I originally suggested in https://github.com/JuliaCloud/AWSS3.jl/pull/182#discussion_r686264312. Is there a reason you didn't go with this?

Reply:
> Is there a reason you didn't go with this?
...because I misread your suggestion. 😶 will go with it now!